Fix compiler warnings and default to -Wall

MichaelEischer opened the pull request:
This fixes all compiler warnings using g++ 5.2.1 (Ubuntu 15.10). Compiling should now also be possible using clang, but that results in a strange linker error for me (probably an incompatibility between g++ 5 and clang related to std::string).

Most changes are unsigned vs signed int, unused variables.

Special ones:
- Allocate marker array in _cmpattern_teamdetector.cpp_ using `new[]`. This allows compilation using clang
- Remove the TeamDetectorInterface from _cmpattern_teamdetector.h_. The interface is not used and defines an update function that differs from the one that's actually used.
- Used isnan from cmath in _glcamera.h_ .
- Fix operators in _gvector.h_ : `+=` would expand to `x = x += p.x;`. g++ complains about possibly undefined behavior. Replaced with `x += p.x`
- In _random.cpp_ : `if (!left < 0)` is always false according to clang. The correct implementation is probably `if (left < 0)

ezavesky commented:
These look sensible.  Changes like [this one](https://github.com/RoboCup-SSL/ssl-vision/pull/53/files#diff-0af35ebc8eb7ce9a79873b5b1432a20aR419) will help developers more easily build the app on other platforms, too.

MichaelEischer commented:
I've rebased the original commit and added a second one that makes the codebase compatible with C++11 while staying compatible with C++98.
@joydeep-b: Any chance to merge this any time soon?

joydeep-b commented:
@MichaelEischer have you tested this with a four-cam firewire setup to verify that the core functionality is correct?
I have a few questions regarding specific line changes, shall comment inline.

joydeep-b commented on src/app/plugins/plugin_dvr.cpp:
```diff
@@ -364,7 +364,6 @@ void PluginDVR::slotMovieSave() {
         if (output.getNumBytes() > 0) {
           //write file:
           QString num = QString::number(i);
-          int n = max(5,num.length());
```
Could you please explain why this is no longer required?

joydeep-b commented on src/app/plugins/plugin_legacypublishgeometry.cpp:
```diff
@@ -48,8 +48,8 @@ PluginLegacyPublishGeometry::PluginLegacyPublishGeometry(
     FrameBuffer* fb,
     RoboCupSSLServer* ds_udp_server_old,
     const RoboCupField& field) :
-    _ds_udp_server_old(ds_udp_server_old),
```
Could you please explain why this line had to be moved?

joydeep-b commented on src/graphicalClient/soccerview.cpp:
```diff
@@ -99,8 +107,7 @@ void GLSoccerView::mousePressEvent(QMouseEvent* event)
 {
-  bool shiftKey = event->modifiers().testFlag(Qt::ShiftModifier);
-  bool ctrlKey = event->modifiers().testFlag(Qt::ControlModifier);
+  (void)event;
```
Does removing this result in an unused variable warning? If so, what compiler and version?

joydeep-b commented on src/shared/capture/capture_generator.cpp:
```diff
@@ -138,7 +138,6 @@ RawImage CaptureGenerator::getFrame()
     int slice_width = w/n_colors;
     rgb color=RGB::Black;
     rgb color2;
-    int gradient=h/256;
```
Is this an unused var?

joydeep-b commented on src/shared/capture/capturev4l.cpp:
```diff
@@ -818,18 +817,12 @@ void CaptureV4L::readParameterValues(VarList * item) {
     for (unsigned int i=0;i<children.size();i++) {
         if (children[i]->getType()==VARTYPE_ID_BOOL && children[i]->getName()=="was_read") vwasread=(VarBool *)children[i];
         if (children[i]->getType()==VARTYPE_ID_BOOL && children[i]->getName()=="enabled") venabled=(VarBool *)children[i];
-        if (children[i]->getType()==VARTYPE_ID_BOOL && children[i]->getName()=="default") vdefault=(VarBool *)children[i];
```
Are these lines removed since the vars are set but not used?

joydeep-b commented on src/shared/capture/capturev4l.cpp:
```diff
@@ -884,18 +877,15 @@ void CaptureV4L::writeParameterValues(VarList * item) {
         if (children[i]->getType()==VARTYPE_ID_BOOL && children[i]->getName()=="enabled") venabled=(VarBool *)children[i];
-        if (children[i]->getType()==VARTYPE_ID_BOOL && children[i]->getName()=="auto") vauto=(VarBool *)children[i];
-        if (children[i]->getType()==VARTYPE_ID_BOOL && children[i]->getName()=="default") vauto=(VarBool *)children[i];
+        if (children[i]->getType()==VARTYPE_ID_BOOL && children[i]->getName()=="default") vdefault=(VarBool *)children[i];
```
Why is "default" written (line 888), bu t not read (line 831)?

MichaelEischer replied:
vdefault is somehow used to reset vint. Not sure what's exactly happening there.

joydeep-b commented on src/shared/capture/capturev4l.cpp:
```diff
@@ -939,23 +929,13 @@ void CaptureV4L::readParameterProperty(VarList * item) {
     vector<VarType *> children=item->getChildren();
     for (unsigned int i=0;i<children.size();i++) {
         if (children[i]->getType()==VARTYPE_ID_BOOL && children[i]->getName()=="enabled") venabled=(VarBool *)children[i];
-        if (children[i]->getType()==VARTYPE_ID_BOOL && children[i]->getName()=="auto") vauto=(VarBool *)children[i];
```
Could you please explain why these lines have been removed?

MichaelEischer replied:
Those variables are once again never read. I've just noticed that these changes should be accompanied by a cleanup of the var type tree.

joydeep-b commented on src/shared/cmpattern/cmpattern_teamdetector.cpp:
```diff
@@ -416,7 +416,7 @@ void TeamDetector::findRobotsByModel(::google::protobuf::RepeatedPtrField< ::SSL
   (void)image;
   const int MaxDetections = _other_markers_max_detections;
   Marker cen; // center marker
-  Marker markers[MaxDetections];
```
Why did this need to be converted from a stack allocation to a heap allocation?

MichaelEischer replied:
clang complains otherwise:
> variable length array of non-POD element type 'CMPattern::Marker'

POD is short for Plain Old Data. The Marker class has a constructor and thus doesn't count as POD.

joydeep-b commented on src/shared/cmpattern/cmpattern_teamdetector.h:
```diff
@@ -40,14 +40,6 @@
 using namespace std;
 namespace CMPattern {
 
-class TeamDetectorInterface {
```
Why did this have to be removed?

MichaelEischer replied:
The update function was out of sync and it's not used anywhere else.

MichaelEischer commented on src/shared/cmpattern/cmpattern_teamdetector.cpp:
```diff
@@ -506,6 +506,7 @@ void TeamDetector::findRobotsByModel(::google::protobuf::RepeatedPtrField< ::SSL
     robots->RemoveLast();
   }
 
+  delete markers;
```
This should be `delete[]` . I'm a bit surprised that I didn't notice this given that clang is complaining now.

MichaelEischer commented on src/shared/util/gvector.h:
```diff
@@ -212,9 +212,9 @@ vector3d<num> cross(const vector3d<num> a,const vector3d<num> b)
   template <class num> \
   vector3d<num> &vector3d<num>::operator opr (const vector3d<num> p) \
   {                  \
-    x = x opr p.x;   \
```
I think it was g++ that complained about these assignments. Took me a while to understand what's happening there.

MichaelEischer commented:
I've successfully tested the changes with a robot shooting goals in a two camera setup. Currently only a few further cleanups of the capturev4l code remain.

joydeep-b commented:
Michael, thanks for this changeset.
Looking through the source, it seems that the v4l drivers are sorely in need of cleanup, along with some paring of the associated VarTree. I'll add that as an issue. 
For now, this changeset looks good to me.
